### PR TITLE
Prevent backend from overwriting orders, customers and articles that have been modified in the meantime

### PR DESCRIPTION
**Pull request by windaishi**
This is a new version of the closed PR #1468, which includes the requested improvements as well as tests.

### 1. Why is this change necessary?

When a backend user saves an entity like `Article`, `Order` or `Customer`, any changes that may have happened concurrently after the entity was opened in the backend up to the point in time where the user presses the "Save" button are overwritten.

Some users tend to open multiple orders, leave them open to do something else and then come back to edit them later. In these cases, write conflicts can and do happen. This may be the case when another backend user has modified the order in the meantime. However, something that we've seen quite often is that orders are modified over the API (e.g. when the order is being picked or shipped using Pickware Mobile), but these changes are then overwritten by another user saving a stale representation of the order from the backend. This causes problems which are very hard to debug and hard to recover without a lot of manual steps.

Because of this, getting this PR merged is a top priority for us (Pickware).

### 2. What does this change do, exactly?

Blocks saving an order, article (item) or customer if the respective entity was modified by someone else after the entity was loaded by the backend. The user gets a message instructing them to close and re-open the entity in question and try again:

peek 2018-01-15 20-25

### 3. Describe each step to reproduce the issue or behaviour.
Open two orders at the same time in the backend, change something, save both. The second save of the changes will completely revert the first changes.

### 4. Please link to the relevant issues (if any).

None.

### 5. Which documentation changes (if any) need to be made because of this PR?

None.

### 6. Checklist

- [X] I have written tests and verified that they fail without my change
- [X] I have squashed any insignificant commits
- [X] This change has comments for package types, values, functions, and non-obvious lines of code
- [X] I have read the contribution requirements and fulfil them.

### CODE REVIEW

File: _sql/migrations/965-add-changed-timestamp-columns.php [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+ * trademark license. Therefore any rights, title and interest in
+ * our trademarks remain entirely with us.
+ */
+
+use Shopware\Components\Migrations\AbstractMigration;
+
+class Migrations_Migration965 extends AbstractMigration
+{
+    /**
+     * Add a column 'changed' in 's_order' that saves the timestamp of the last change of the order.
```
Comment (fixpunkt) [inline _sql/migrations/965-add-changed-timestamp-columns.php]: Comment is out of date.

File: engine/Shopware/Controllers/Backend/Article.php [hunk 2]
```diff
@@ -157,7 +157,25 @@ public function saveAction()
...
         if ($this->Request()->has('id')) {
+            /** @var Article $article */
             $article = $this->getRepository()->find((int) $this->Request()->getParam('id'));
+            // Check whether the article has been modified in the meantime
+            try {
+                $data['changed'] = new \DateTime($data['changed']);
+            } catch (\Exception $e) {
+                $data['changed'] = null;
+            }
+            if ($data['changed'] === null || $article->getChanged() != $data['changed']) {
```
Comment (fixpunkt) [inline engine/Shopware/Controllers/Backend/Article.php]: In which situation is `$data['changed'] === null`? Shouldn't we skip the check in this case?

File: engine/Shopware/Controllers/Backend/Article.php [hunk 3]
```diff
@@ -157,7 +157,25 @@ public function saveAction()
...
+                $namespace = Shopware()->Snippets()->getNamespace('backend/article/controller/main');
+
+                $this->View()->assign([
+                    'success' => false,
+                    'data' => $this->getArticle($article->getId()),
```
Comment (fixpunkt) [inline engine/Shopware/Controllers/Backend/Article.php:168]: Why do you return the article?

File: engine/Shopware/Controllers/Backend/Article.php [hunk 4]
```diff
@@ -3583,7 +3601,7 @@ protected function prepareArticleAssociatedData($data)
             $data['propertyGroup'] = null;
         }
 
-        $data['changed'] = new \DateTime();
+        unset($data['changed']);
```
Comment (fixpunkt) [inline engine/Shopware/Controllers/Backend/Article.php]: This line needs a comment

File: engine/Shopware/Controllers/Backend/Article.php [hunk 5]
```diff
@@ -157,7 +157,25 @@ public function saveAction()
     {
         $data = $this->Request()->getParams();
         if ($this->Request()->has('id')) {
+            /** @var Article $article */
             $article = $this->getRepository()->find((int) $this->Request()->getParam('id'));
+            // Check whether the article has been modified in the meantime
+            try {
+                $data['changed'] = new \DateTime($data['changed']);
```
Comment (fixpunkt) [inline engine/Shopware/Controllers/Backend/Article.php]: I think you shouldn't set `$data['changed']` here and use a local variable instead, because that has less side effects - since `$data['changed']` is cleared anyway before saving.

File: engine/Shopware/Controllers/Backend/Article.php (hunk 2 above)
Comment (fixpunkt) [reply engine/Shopware/Controllers/Backend/Article.php]: Ok I see, only on exception? We need to untangle the control flow a bit here...

File: engine/Shopware/Controllers/Backend/Customer.php [hunk 6]
```diff
@@ -296,6 +296,24 @@ public function saveAction()
 
         $params = $this->Request()->getParams();
 
+        // Check whether the customer has been modified in the meantime
+        try {
+            $params['changed'] = new \DateTime($params['changed']);
```
Comment (fixpunkt) [inline engine/Shopware/Controllers/Backend/Customer.php]: The same comments as for `Article` apply here too.

File: engine/Shopware/Controllers/Backend/Order.php [hunk 7]
```diff
@@ -407,6 +407,22 @@ public function saveAction()
         //get all passed order data
         $data = $this->Request()->getParams();
 
+        // Check whether the order has been modified in the meantime
+        try {
+            $data['changed'] = new \DateTime($data['changed']);
```
Comment (fixpunkt) [inline engine/Shopware/Controllers/Backend/Order.php]: Same

File: _sql/migrations/965-add-changed-timestamp-columns.php [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+     *
+     * @param string $modus
+     */
+    public function up($modus)
+    {
+        $sql = <<<'EOD'
+ALTER TABLE s_order ADD COLUMN changed DATETIME NOT NULL DEFAULT CURRENT_TIMESTAMP() AFTER id;
```
Comment (fixpunkt) [inline _sql/migrations/965-add-changed-timestamp-columns.php]: We should probably remove the `AFTER id` part.

File: _sql/migrations/965-add-changed-timestamp-columns.php [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+ALTER TABLE s_user ADD COLUMN changed DATETIME NOT NULL DEFAULT CURRENT_TIMESTAMP() AFTER id;
```
Comment (fixpunkt) [inline _sql/migrations/965-add-changed-timestamp-columns.php]: We should probably remove the `AFTER id` part.

File: themes/Backend/ExtJs/backend/order/model/order.js [hunk 10]
```diff
@@ -162,7 +163,8 @@ Ext.define('Shopware.apps.Order.model.Order', {
          */
         reader: {
             type: 'json',
-            root: 'data'
+            root: 'data',
+            messageProperty: 'message',
```
Comment (fixpunkt) [inline themes/Backend/ExtJs/backend/order/model/order.js]: We need to be consistent with Shopware's code style here, so please remove the trailing comma.

**Review by fixpunkt — CHANGES_REQUESTED** [review]
Nice work!

We need to update the description to better explain the fact that this PR is not only about write conflicts caused by multiple backend users modifying the same entity at the same time (which should be rare), but write conflicts between API accesses or, in the case of article, frontend sales (is my thinking correct here?), which really sucks.

File: engine/Shopware/Controllers/Backend/Article.php (hunk 2 above)
Comment (windaishi) [reply engine/Shopware/Controllers/Backend/Article.php]: Suggestions?

File: engine/Shopware/Controllers/Backend/Article.php (hunk 3 above)
Comment (windaishi) [reply engine/Shopware/Controllers/Backend/Article.php:168]: Showare does that at every place

File: engine/Shopware/Controllers/Backend/Article.php (hunk 4 above)
Comment (windaishi) [reply engine/Shopware/Controllers/Backend/Article.php]: Ah, ya, maybe

**Comment by windaishi** [comment]
I did the requested changes.

File: snippets/backend/article/controller/main.ini [hunk 11]
```diff
@@ -0,0 +1,5 @@
+[en_GB]
+article_has_been_changed = "The article has been changed in the meantime. To prevent overwriting these changes, saving the article was aborted. Please close the article and re-open it."
```
Comment (fixpunkt) [inline snippets/backend/article/controller/main.ini]: The actual text should use "item" instead of "article", since article is Ginglish for item (or product). Keep the snippet name though.

**Review by fixpunkt — CHANGES_REQUESTED** [review]
Nice work!

Didn't we also decide we wanted tests for this PR?

**Review by fixpunkt — APPROVED** [review]

**Comment by fixpunkt** [comment]
New PR is here: https://github.com/shopware/shopware/pull/1474

@maximilianoster found a relevant corner case. Steps to reproduce:

1. Open the order list.
2. Open an order from the order list (window A).
3. Open the same order another time from the order list (window B).
4. Change something in window A, save and close window A.
5. Change something else in window B, save and close window B.
6. Re-open the order. The changes from 4 were overwritten by the changes from 5.

The problem here is probably due to the fact that we only compare the `changed` timestamp in the model, but the form state is stale anyway. We should instead add a field to the form which has the `changed` timestamp as its value and which is then sent to the controller action. While we're at it, we might as well expose this to the user.

**Comment by fixpunkt** [comment]
@samuelvogel just reminded me that the `changed` timestamp is already shown for products:

product modified timestamp

So we should use a similar presentation for orders and customers. Since this is a form-group title, we'll probably still need to add a hidden input, too.

**Comment by fixpunkt** [comment]
We've also just discovered that the monkey patch doesn't allow you to change the order status twice (or order status, then payment status) from the order list. This is something we need to test for the PR too!

**Comment by windaishi** [comment]
Should be fixed now

**Comment by fixpunkt** [comment]
We have another bug in the monkey patch: when an order position is added or updated with a price change that causes the order total to change, subsequent save operations on the order fail, too.

Can you test whether this PR is also affected by that?

**Comment by windaishi** [comment]
I have fixed several problem with the overwrite protection. The PR is updated already. @maximilianoster still searches for more bugs.

**Comment by svenmuennich** [comment]
Can we close this PR? shopware/shopware#1474 was opened (even though it's currently failing) and should be updated as we continue to find more edge cases.

@fixpunkt Please close this PR if you agree.
